File: src/day1/BinarySearchList.go
package main

import "fmt"

func bsList(haystack []int, needle int) bool {
	low := 0
	high := len(haystack) - 1

	for low <= high {
		middle := (low + high) / 2

		if haystack[middle] == needle {
			return true
		} else if haystack[middle] < needle {
			low = middle + 1
		} else {
			high = middle - 1
		}
	}
	return false
}

func main() {
	foo := []int{1, 3, 4, 69, 71, 81, 90, 99, 420, 1337, 69420}
	fmt.Println(bsList(foo, 69))
	fmt.Println(bsList(foo, 1336))
	fmt.Println(bsList(foo, 69420))
	fmt.Println(bsList(foo, 69421))
	fmt.Println(bsList(foo, 1))
	fmt.Println(bsList(foo, 0))
}

File: src/day1/BinarySearchList.ts
export default function bs_list(haystack: number[], needle: number): boolean {
    let low :number = 0;
    let high :number = haystack.length;

    do {
        const middle :number = Math.floor(low + (high - low) / 2);
        const value :number = haystack[middle];

        if (value === needle) {
            return true;
        } else if (value > needle) {
            high = middle;
        } else {
            low = middle + 1;
        }
    } while (low < high);

    return false;
}
